File: python/custom_icon_theme_and_fonts.py
########################################################################################
# This file is licensed under CC0:                                                     #
#                                                                                      #
#      https://github.com/svenevs/nanogui-customization-demo/blob/master/LICENSE       #
########################################################################################

import gc
import nanogui
from nanogui import entypo, fontawesome

import os
import sys
sys.path.insert(0, os.path.abspath(os.path.dirname(__file__)))
from custom_screen import CustomScreen
from custom_theme import ThemeChoice
from make_compare_window import makeCompareWindow


if __name__ == "__main__":
    nanogui.init()
    screen = CustomScreen((800, 900))

    # manual demonstration
    window = nanogui.Window(screen, "Manual Labels")
    window.setLayout(nanogui.GridLayout())
    nanogui.Label(window, "Entypo+")
    nanogui.Label(window, "Awesome")

    # just some similar icons to compare with
    icons = [
        (        entypo.ICON_INFO_WITH_CIRCLE,    fontawesome.ICON_SOLID_INFO_CIRCLE),
        (                    entypo.ICON_HOME,           fontawesome.ICON_SOLID_HOME),
        (              entypo.ICON_LIGHT_BULB,            fontawesome.ICON_LIGHTBULB),
        (                    entypo.ICON_LINK,           fontawesome.ICON_SOLID_LINK),
        (                     entypo.ICON_EYE,            fontawesome.ICON_SOLID_EYE),
        (           entypo.ICON_EYE_WITH_LINE,      fontawesome.ICON_SOLID_EYE_SLASH),
        (                 entypo.ICON_COMPASS,              fontawesome.ICON_COMPASS),
        (entypo.ICON_CONTROLLER_FAST_BACKWARD,  fontawesome.ICON_SOLID_FAST_BACKWARD),
        ( entypo.ICON_CONTROLLER_FAST_FORWARD,   fontawesome.ICON_SOLID_FAST_FORWARD),
        (entypo.ICON_CONTROLLER_JUMP_TO_START,  fontawesome.ICON_SOLID_STEP_BACKWARD),
        (         entypo.ICON_CONTROLLER_NEXT,   fontawesome.ICON_SOLID_STEP_FORWARD),
        (         entypo.ICON_CONTROLLER_PAUS,          fontawesome.ICON_SOLID_PAUSE),
        (         entypo.ICON_CONTROLLER_PLAY,           fontawesome.ICON_SOLID_PLAY),
        (       entypo.ICON_CONTROLLER_RECORD,         fontawesome.ICON_SOLID_CIRCLE),
        (         entypo.ICON_CONTROLLER_STOP,           fontawesome.ICON_SOLID_STOP),
        (                    entypo.ICON_CROP,           fontawesome.ICON_SOLID_CROP),
        (                entypo.ICON_DOWNLOAD,       fontawesome.ICON_SOLID_DOWNLOAD),
        (               entypo.ICON_CLIPBOARD,      fontawesome.ICON_SOLID_CLIPBOARD),
        (                   entypo.ICON_CLOCK,                fontawesome.ICON_CLOCK),
        (               entypo.ICON_THUMBS_UP,      fontawesome.ICON_SOLID_THUMBS_UP),
        (             entypo.ICON_THUMBS_DOWN,    fontawesome.ICON_SOLID_THUMBS_DOWN),
        (             entypo.ICON_PAPER_PLANE,    fontawesome.ICON_SOLID_PAPER_PLANE),
        (                  entypo.ICON_ROCKET,         fontawesome.ICON_SOLID_ROCKET),
        (                     entypo.ICON_KEY,            fontawesome.ICON_SOLID_KEY),
        (                   entypo.ICON_GAUGE, fontawesome.ICON_SOLID_TACHOMETER_ALT)
    ]
    for entypo_icon, fontawesome_icon in icons:
        # add some vertical spacing (two widgets per row)
        nanogui.Label(window, " ")
        nanogui.Label(window, " ")
        # add the entypo icon
        nanogui.Label(window, nanogui.utf8(entypo_icon), "icons")
        # add the fontawesome icon
        nanogui.Label(window, nanogui.utf8(fontawesome_icon), "fontawesome")

    # add some vertical spacing (two widgets per row)
    nanogui.Label(window, " ")
    nanogui.Label(window, " ")
    window.setFixedWidth(125)

    # make one with the default theme
    window = makeCompareWindow(screen, "Default Theme", ThemeChoice.Default)
    window.setPosition((150, 0))

    # make one with the custom theme
    window = makeCompareWindow(screen, "Fontawesome Icon Theme", ThemeChoice.Fontawesome)
    window.setPosition((475, 0))

    screen.performLayout()
    screen.drawAll()
    screen.setVisible(True)
    nanogui.mainloop()
    del screen
    gc.collect()
    nanogui.shutdown()

File: cpp/custom_icon_theme_and_fonts.cpp
/***************************************************************************************
 * This file is licensed under CC0:                                                    *
 *                                                                                     *
 *      https://github.com/svenevs/nanogui-customization-demo/blob/master/LICENSE      *
 **************************************************************************************/
#include "make_compare_window.hpp" // includes custom_screen and custom_theme

#include <nanogui/nanogui.h>

#include <vector>

int main(void) {
    nanogui::init();

    {
        using namespace nanogui;

        CustomScreen *screen = new CustomScreen({800, 900});

        // manual demonstration
        Window *window = new Window(screen, "Manual Labels");
        window->setLayout(new GridLayout());
        new Label(window, "Entypo+");
        new Label(window, "Awesome");

        // just some similar icons to compare with
        const std::vector<std::pair<int, int>> icons = {
            {        ENTYPO_ICON_INFO_WITH_CIRCLE,    FONTAWESOME_ICON_SOLID_INFO_CIRCLE},
            {                    ENTYPO_ICON_HOME,           FONTAWESOME_ICON_SOLID_HOME},
            {              ENTYPO_ICON_LIGHT_BULB,            FONTAWESOME_ICON_LIGHTBULB},
            {                    ENTYPO_ICON_LINK,           FONTAWESOME_ICON_SOLID_LINK},
            {                     ENTYPO_ICON_EYE,            FONTAWESOME_ICON_SOLID_EYE},
            {           ENTYPO_ICON_EYE_WITH_LINE,      FONTAWESOME_ICON_SOLID_EYE_SLASH},
            {                 ENTYPO_ICON_COMPASS,              FONTAWESOME_ICON_COMPASS},
            {ENTYPO_ICON_CONTROLLER_FAST_BACKWARD,  FONTAWESOME_ICON_SOLID_FAST_BACKWARD},
            { ENTYPO_ICON_CONTROLLER_FAST_FORWARD,   FONTAWESOME_ICON_SOLID_FAST_FORWARD},
            {ENTYPO_ICON_CONTROLLER_JUMP_TO_START,  FONTAWESOME_ICON_SOLID_STEP_BACKWARD},
            {         ENTYPO_ICON_CONTROLLER_NEXT,   FONTAWESOME_ICON_SOLID_STEP_FORWARD},
            {         ENTYPO_ICON_CONTROLLER_PAUS,          FONTAWESOME_ICON_SOLID_PAUSE},
            {         ENTYPO_ICON_CONTROLLER_PLAY,           FONTAWESOME_ICON_SOLID_PLAY},
            {       ENTYPO_ICON_CONTROLLER_RECORD,         FONTAWESOME_ICON_SOLID_CIRCLE},
            {         ENTYPO_ICON_CONTROLLER_STOP,           FONTAWESOME_ICON_SOLID_STOP},
            {                    ENTYPO_ICON_CROP,           FONTAWESOME_ICON_SOLID_CROP},
            {                ENTYPO_ICON_DOWNLOAD,       FONTAWESOME_ICON_SOLID_DOWNLOAD},
            {               ENTYPO_ICON_CLIPBOARD,      FONTAWESOME_ICON_SOLID_CLIPBOARD},
            {                   ENTYPO_ICON_CLOCK,                FONTAWESOME_ICON_CLOCK},
            {               ENTYPO_ICON_THUMBS_UP,      FONTAWESOME_ICON_SOLID_THUMBS_UP},
            {             ENTYPO_ICON_THUMBS_DOWN,    FONTAWESOME_ICON_SOLID_THUMBS_DOWN},
            {             ENTYPO_ICON_PAPER_PLANE,    FONTAWESOME_ICON_SOLID_PAPER_PLANE},
            {                  ENTYPO_ICON_ROCKET,         FONTAWESOME_ICON_SOLID_ROCKET},
            {                     ENTYPO_ICON_KEY,            FONTAWESOME_ICON_SOLID_KEY},
            {                   ENTYPO_ICON_GAUGE, FONTAWESOME_ICON_SOLID_TACHOMETER_ALT}
        };
        for (const auto &pair : icons) {
            // add some vertical spacing (two widgets per row)
            new Label(window, " "); new Label(window, " ");
            // add the entypo icon
            new Label(window, std::string(nanogui::utf8(pair.first).data()), "icons");
            // add the fontawesome icon
            new Label(window, std::string(nanogui::utf8(pair.second).data()), "fontawesome");
        }
        // add some vertical spacing (two widgets per row)
        new Label(window, " "); new Label(window, " ");
        window->setFixedWidth(125);

        // make one with the default theme
        window = makeCompareWindow(screen, "Default Theme", ThemeChoice::Default);
        window->setPosition({150, 0});

        // make one with the custom theme
        window = makeCompareWindow(screen, "Fontawesome Icon Theme", ThemeChoice::Fontawesome);
        window->setPosition({475, 0});

        screen->setVisible(true);
        screen->performLayout();

        nanogui::mainloop();

    }

    nanogui::shutdown();


    return 0;
}
